File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] m) {
        for(int i=0;i<m.length;i++){
            for(int j=0;j<i;j++){
                int temp=m[i][j];
                m[i][j]=m[j][i];
                m[j][i]=temp;
            }
        }
        for(int i=0;i<m.length;i++){
            int stcol=0;
            int endcol=m[0].length-1;
            while(stcol<endcol){
                int temp=m[i][stcol];
                m[i][stcol]=m[i][endcol];
                m[i][endcol]=temp;
                stcol++;
                endcol--;
            }
        }
    }
}

File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // firstly transpose matrix

        int n=matrix.size();
        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                swap(matrix[i][j],matrix[j][i]);
            }
        }

            // then each row

        for(int i=0;i<n;i++){
            int st=0,end=n-1;
            while(st<end){
                swap(matrix[i][st],matrix[i][end]);
                st++;
                end--;
            }
        }

        }
};